File: KVSRITHTML/KvsritHtmlKVersion/download_imgs.py
import os
import re
import requests
from urllib.parse import urljoin, urlparse, unquote

url = "https://drkvsrit.ac.in/"
dest_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "images", "drkvsrit")
os.makedirs(dest_dir, exist_ok=True)

requests.packages.urllib3.disable_warnings()
session = requests.Session()
session.verify = False
session.headers.update({
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
})

def ensure_absolute(img_url):
    if img_url.startswith("//"):
        return "https:" + img_url
    if img_url.startswith("/"):
        return urljoin(url, img_url)
    if img_url.startswith("http"):
        return img_url
    return urljoin(url, img_url)

print("Fetching website...")
try:
    response = session.get(url, timeout=30)
    html = response.text
except Exception as e:
    print(f"Error fetching website: {e}")
    exit(1)

# Regex to find <img src="...">
img_matches = set(re.findall(r'<img[^>]+src=[\'"]([^\'"]+)[\'"]', html, re.IGNORECASE))
# Regex to find background-image: url(...)
bg_matches = set(re.findall(r'url\([\'"]?([^\'"\)]+)[\'"]?\)', html, re.IGNORECASE))

all_images = []
for m in img_matches.union(bg_matches):
    if m.startswith('data:'):
        continue
    img_url = ensure_absolute(m)
    if img_url not in all_images:
        all_images.append(img_url)

print(f"Found {len(all_images)} image URLs...")

for i, img_url in enumerate(all_images):
    print(f"Downloading {i+1}/{len(all_images)}: {img_url}")
    parsed = urlparse(img_url)
    filename = os.path.basename(unquote(parsed.path)) or f"image_{i}.jpg"
    if not re.search(r'\.(jpg|jpeg|png|gif|webp|svg|ico)$', filename, re.IGNORECASE):
        filename += '.jpg'
    
    base, ext = os.path.splitext(filename)
    final_dest = os.path.join(dest_dir, filename)
    counter = 1
    while os.path.exists(final_dest):
        final_dest = os.path.join(dest_dir, f"{base}_{counter}{ext}")
        counter += 1

    try:
        img_res = session.get(img_url, timeout=30)
        img_res.raise_for_status()
        with open(final_dest, 'wb') as f:
            f.write(img_res.content)
    except Exception as e:
        print(f"Failed to download {img_url}: {e}")

print("Done downloading images.")

File: KVSRITHTML/KvsritHtmlKVersion/download_imgs.js
const fs = require('fs');
const path = require('path');

const websiteUrl = 'https://drkvsrit.ac.in/';
const destDir = path.join(__dirname, 'images', 'drkvsrit');

if (!fs.existsSync(destDir)) {
    fs.mkdirSync(destDir, { recursive: true });
}

function ensureUrlAbsolute(urlStr) {
    if (urlStr.startsWith('http://') || urlStr.startsWith('https://')) {
        return urlStr;
    }
    if (urlStr.startsWith('//')) {
        return 'https:' + urlStr;
    }
    if (urlStr.startsWith('/')) {
        return 'https://drkvsrit.ac.in' + urlStr;
    }
    return 'https://drkvsrit.ac.in/' + urlStr;
}

async function scrapeImages() {
    console.log("Fetching website...");
    const res = await fetch(websiteUrl, {
        headers: {
            'User-Agent': 'Mozilla/5.0'
        }
    });
    const data = await res.text();

    const imageUrls = [];

    // Match <img src="...">
    const imgRegex = /<img[^>]+src=["']([^"']+)["']/g;
    let match;
    while ((match = imgRegex.exec(data)) !== null) {
        let imgUrl = ensureUrlAbsolute(match[1]);
        if (!imageUrls.includes(imgUrl)) {
            imageUrls.push(imgUrl);
        }
    }

    // Match background-image: url(...)
    const bgRegex = /url\(["']?([^"'\)]+)["']?\)/g;
    while ((match = bgRegex.exec(data)) !== null) {
        let bgUrl = match[1];
        if (bgUrl.startsWith('data:')) continue;
        let imgUrl = ensureUrlAbsolute(bgUrl);
        if (!imageUrls.includes(imgUrl)) {
            imageUrls.push(imgUrl);
        }
    }

    console.log(`Found ${imageUrls.length} image URLs...`);
    for (const [i, rawUrl] of imageUrls.entries()) {
        const imgUrl = encodeURI(rawUrl);
        try {
            let p = new URL(imgUrl).pathname;
            let fileName = path.basename(decodeURIComponent(p)) || `image_${i}.jpg`;
            if (!/\.(jpg|jpeg|png|gif|webp|svg)$/i.test(fileName)) {
                fileName += '.jpg';
            }
            let finalDest = path.join(destDir, fileName);
            let counter = 1;
            while (fs.existsSync(finalDest)) {
                const ext = path.extname(fileName);
                const base = path.basename(fileName, ext);
                finalDest = path.join(destDir, `${base}_${counter}${ext}`);
                counter++;
            }

            console.log(`Downloading ${i + 1}/${imageUrls.length}: ${imgUrl}`);
            const imgRes = await fetch(imgUrl, {
                headers: { 'User-Agent': 'Mozilla/5.0' }
            });
            if (!imgRes.ok) {
                console.error(`Status ${imgRes.status} for ${imgUrl}`);
                continue;
            }
            const buffer = await imgRes.arrayBuffer();
            fs.writeFileSync(finalDest, Buffer.from(buffer));
        } catch (e) {
            console.error(`Error downloading ${imgUrl}:`, e.message);
        }
    }
    console.log("Done downloading images.");
}

scrapeImages().catch(console.error);
